Web console: display and attach persistent volume claims

fabianofranz opened the pull request:
~~Code is **not yet ready for review** but I'm opening this for design feedback.~~ PTAL @jwforres @ajacobs21e @spadgett @benjaminapetersen 

List all existing persistent volume claims, this is under *Browse/Storage*. The "Status" column represents the fulfilment of a claim against actual volumes in the cluster. You can click a claim to see the detail page with a little more information.

~~[deprecated screenshot](https://cloud.githubusercontent.com/assets/158611/11859716/eefae26c-a454-11e5-83e0-072ce7021c14.png)~~ (scroll down for the most up-to-date)

Now how do we attach an existing persistent volume claim to pods? Volumes allow you to attach a pvc to *any resource that declares a pod spec*, so deployment configs, deployments (replication controllers in practice), etc. It does not seem possible to attach to a pod directly, and attaching to a deployment does not seem to make much sense (will not cause a new deployment so I think the pods would only be affected if a redeployment occurs). So I'm only supporting attaching to deployment configs at the moment, but other kinds can be added easily. 

I'm doing this in the deployment config details page using the *Attach persistent volume claim* link (similar to how it's done in *Create route*). Volumes were not yet present in the deployment config details page, so I added it mostly with the same design it's displayed in pod details, but together with the "Template" section (since it's not volumes *of the deployment config*, but of the template that will result in the dc's pods).

~~[deprecated screenshot](https://cloud.githubusercontent.com/assets/158611/11859720/f73ae314-a454-11e5-9d83-2a7474163430.png)~~ (scroll down for the most up-to-date)

Here is where you attach an existing persistent volume claim to a deployment config. I'm not particularly happy with the dropdown for the pvcs, it would be interesting to have more information (probably capacity, status and access modes) instead of just the name, so that it's more clear to users what's the pvc they are attaching to. What would you suggest, maybe radios? Also, it's still missing a few simple inputs (mount path, volume name, etc).

~~[deprecated screenshot](https://cloud.githubusercontent.com/assets/158611/11859721/fcc58bf4-a454-11e5-92b4-11e1b4eb155b.png)~~

jwforres commented:
"Attach existing persistent volume claim" is a really long action link, maybe just drop the existing

jwforres commented:
We should update the Volumes section on the pod page when a volume is a PVC to link to the PVC page, like you've done on the dep config page.

jwforres commented:
we should probably let you attach PVCs to a replication controller if its not owned by a deployment config.  a scale up/down (or deleting the old pods) would create new pods and attach the volumes right?

jwforres commented:
Agree that the attach page should have more information about the PVCs the user is choosing from, definitely access modes and capacity.  One possibility would be a table with a radio button on the left of each row, think like the UI you see for selecting instance type after you go to launch an instance in AWS (or for selecting the security group).

jwforres commented:
From the page were you are looking at a particular volume claim, you probably want the "Attach this to..." action

fabianofranz commented:
@jwforres Thanks for the comments! I've been back and forth about "Attach this to..." from the pvc details because of how you would select the resource to attach to - they can be of multiple kinds, so what kind of ui control would we use? Two dependent selects (first kind then load names on the second select) maybe?

jwforres commented:
Yeah so far we try not to surface replication controllers as a concept
unless you actually have some, so that would make this tricky...
On Dec 17, 2015 8:49 AM, "Fabiano Franz" <notifications@github.com> wrote:

> @jwforres <https://github.com/jwforres> Thanks for the comments! I've
> been back and forth about "Attach this to..." from the pvc details because
> of how you would select the resource to attach to - they can be of multiple
> kinds, so what kind of ui control would we use? Two dependent selects
> (first kind then load names on the second select) maybe?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6358#issuecomment-165457529>.
>

fabianofranz commented:
Something I still have to define (and it has to do with what @smarterclayton was saying about the life cycle of a volume attachment, if it survives to pod creation/destruction) is to which resources we will allow attaching existing pvc's. 

Right now I'm *only* allowing attach to deployment configs, so it's volumes that are restored when new deployments happen. I *do* display the "Attach persistent volume claim" link in the pod details page, but I point it to the deployment config that originated the pods, so you are always attaching to a dc. So, question is: should we also allow attaching directly to deployments? What about other regular replication controllers (in case we display them on the web console, I'm not sure), or any other resources?

benjaminapetersen commented on assets/app/scripts/app.js:
```diff
@@ -194,6 +195,14 @@ angular
+        templateUrl: 'views/storage.html',
+        controller: 'StorageController'
+      })
+      .when('/project/:project/browse/persistentvolumeclaims/:pvc', {
```
I know you mentioned code not ready, but figured I'd throw a thought or two in here. Would we want to keep these routes associated?

```javascript
.when('/project/:project/browse/storage'
.when('/project/:project/browse/storage/persistentvolumeclaims/:pvc', {
.when('/project/:project/browse/storage/persistentvolumeclaims/attach', {  
```
As we add routes I think it will pay off if we can keep patterns as consistent as possible.

liggitt replied:
what if you have a pvc named `attach`?

benjaminapetersen replied:
Then you will be very sad about your name decision :smile:

benjaminapetersen replied:
Great Q, what i suggested is prob not the best.  Is `.when('/project/:project/attachPersistentVolumeClaim')` most fitting?  Im assuming with this path it would be being attached to the project, rather than a pod/deployment/etc.

spadgett replied:
Wondering if it's too late to use short names in our URLs: `/browse/pvc`, `/attachPVC`? Not consistent with what we already have unfortunately.

fabianofranz replied:
The inspiration for `/project/:project/attachPersistentVolumeClaim` was `/project/:project/createRoute`, but I'm open to change it to what we think fits best. Another idea would be to place it under the resource (pretty long though), like

```
/project/:project/browse/deployments/:deploymentconfig/attach/pvc
```

benjaminapetersen replied:
I do actually prefer that option @fabianofranz ...cleaner, and seems like it will help us continue to add urls in an organized way in the future.

spadgett replied:
Concerned with that pattern we might run into conflicts with action names and resource names, even if it reads better.

benjaminapetersen replied:
Yeah, `attach-pvc` may be better & avoid that problem?  Def prefer the direction of `/lowercase-hyphenated/`urls, with avoiding any conflicts.

benjaminapetersen commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,71 @@
+            $scope.attachment.persistentVolumeClaim = $scope.pvcs[0];
+          }
+        });
+
```
The logic here looks nice and clean!  I'm inclined to suggest we have a `PersistenVolumeClaim` service & push some of this logic up there & out of the controller. 

```javascript
.factory('PersistentVolumeClaim', [
   function() {
      return {
           attach: function(volume) {
             // ...
             deploymentConfig.spec.template.spec.volumes.push(newVolume);
             // return the promise, so the controller can .then() and handle $scope.alerts stuff
             return DataService.update("deploymentconfigs", deploymentConfig.metadata.name, deploymentConfig, context)
           }
      }
   }
]);
```

fabianofranz replied:
Totally! Moving several bits from here to a service is something I want to before removing WIP.

ajacobs21e commented:
Is the process to bind a claim to a volume and then attach to a pod?

> this is under Browse/Storage

This is outside of this conversation but with routes and storage we are at the point that we have to think about a better way to organize things under "browse" - either 2+ top level nav items that better explain the relationship between sub-items or groupings within the "browse" category. 

>  I added it mostly with the same design it's displayed in pod details, but together with the "Template" section (since it's not volumes of the deployment config, but of the template that will result in the dc's pods).

In that case it should not use the same section header style and should be nested under the template heading. 

>  I'm not particularly happy with the dropdown for the pvcs, it would be interesting to have more information

This would be a good use case for a data list with radio buttons... like Jessica said.

> From the page were you are looking at a particular volume claim, you probably want the "Attach this to..." action

Agree. I can come up with ideas if I understand the use case better. A search scenario with results in a table / data list that shows resource type and maybe sorts by most common resource type. So if a pod is the most common thing you'd want to attach, list pods first, then lower down RCs, etc.

fabianofranz commented:
![attach-pvc-with-table](https://cloud.githubusercontent.com/assets/158611/11882500/df768446-a4f3-11e5-8b2a-7e09907bb463.png)

spadgett commented:
How large can the list of claims get?

Maybe incorporate the Patternfly list view styles?

https://www.patternfly.org/patterns/list-view/

spadgett commented:
How about just Browse -> Persistent Volume Claims and use menu sections labels when we add other storage options?  "Storage - Persistent Volume Claims" is pretty wordy.

https://www.patternfly.org/patterns/vertical-with-persistent-secondary/

fabianofranz commented:
@ajacobs21e 
> Is the process to bind a claim to a volume and then attach to a pod?

Claims are bound to volumes automatically, you just create the claim (I want 2G, read-only, etc) and then the cluster does the binding. Creating a claim is out of scope of this PR though. After having claims then you attach them to a resource, usually not directly to a pod but to a deployment config for example (which will then make it available to the pods when deploying).

> This would be a good use case for a data list with radio buttons... like Jessica said.

I'm using a table now, take a look at the latest screen above. @spadgett's idea of using a list view also seems to fit, it could be the case to incorporate list view styles if we don't have it already

fabianofranz commented:
I did a number of improvements to the "Attach Storage" page. In terms of features, I added the "Containers" section that allows you to select the container(s) you want to attach and mount the volume to (takes all of them by default, disable the option if you only have one). The CLI does the same through the `--containers` flag (also `*` by default). @bparees I think you mentioned this yesterday.

![attach-one-container](https://cloud.githubusercontent.com/assets/158611/11906935/5a54024c-a5b7-11e5-89e7-7f4b6c1c99cd.png)

![attach-select-container](https://cloud.githubusercontent.com/assets/158611/11906940/5dbdf1d6-a5b7-11e5-916e-670ceaa880ce.png)

jwforres commented:
the "select the persistent volume claim to attach" text could be completely
off screen if there were a lot of volumes in that list, probably needs to
be above the table

i'd like to see the storage size in a scannable column, same with access
mode, if i know i'm looking for my 500GB volume that there is only one of,
its easier if i can scan that list of sizes

for the containers, if you check "all containers" does that select the
checkboxes next to all the containers? like a "Select all" behavior.  im
not in love with the gap between the all containers and the list of
containers, maybe it just needs a gray divider line instead

the first screenshot when there is only one container, something feels off
with the wording but i cant decide whats bothering me about it yet

On Fri, Dec 18, 2015 at 3:46 PM, Fabiano Franz <notifications@github.com>
wrote:

> I did a number of improvements to the "Attach Storage" page. In terms of
> features, I added the "Containers" section that allows you to select the
> container(s) you want to attach and mount the volume to (takes all of them
> by default, disable the option if you only have one). The CLI does the same
> through the --containers flag (also * by default). @bparees
> <https://github.com/bparees> I think you mentioned this yesterday.
>
> [image: attach-one-container]
> <https://cloud.githubusercontent.com/assets/158611/11906935/5a54024c-a5b7-11e5-89e7-7f4b6c1c99cd.png>
>
> [image: attach-select-container]
> <https://cloud.githubusercontent.com/assets/158611/11906940/5dbdf1d6-a5b7-11e5-916e-670ceaa880ce.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6358#issuecomment-165894319>.
>

fabianofranz commented:
@jwforres 
In the first screen the template only declares one container, so I thought I should keep it in the screen so you know what is going to happen, but another idea is to completely remove the "Containers" section if it only declares one. Or just list the one container you have, selected (and disabled) already.

In the second screen, the "all containers" checkbox is selected by default and the containers list is not shown. If you de-select "all containers" then they are shown and you are given the option to choose. 

Not 100% about this though. Maybe a dumber list of the containers, all selected by default, and an option to "select all" would be simpler.

fabianofranz commented:
Updated "Attach" page. I changed the containers section to use the "Show options ..." pattern - when the link is clicked, shows the containers to select individually. Looks much better.

![attach-one-container](https://cloud.githubusercontent.com/assets/158611/11910783/95fddc30-a5e6-11e5-8dd2-815304307c4f.png)

![attach-select-container](https://cloud.githubusercontent.com/assets/158611/11910784/99615848-a5e6-11e5-91d3-295178ce595d.png)

liggitt commented:
Tell them what will happen if they don't click the customize link. If there's only one container, just include the text that tells them the volume will be mounted to that container and omit the link. If there are multiple containers, say "The volume will be mounted into all containers. [Select specific containers](#)" or something

liggitt commented:
do we use "e.g." in other placeholders? looks a little odd

fabianofranz commented:
@liggitt Sounds good. For the placeholders I could just use */data* in Mount path (although it could give the idea that it would use "/data" if not provided), and *(generated if empty)* in Volume name, like we do in template parameters.

fabianofranz commented:
I'll just make it clear in the Volume name help label: "Mount path for the volume inside the container. If not specified, the volume will not be mounted automatically."

fabianofranz commented:
![attach-one-container](https://cloud.githubusercontent.com/assets/158611/11911037/beffb456-a5ef-11e5-8b6f-dd312315e535.png)

![attach-select-container](https://cloud.githubusercontent.com/assets/158611/11911038/c72aed9e-a5ef-11e5-80b1-17a991038d53.png)

liggitt commented:
now that I see it, it's probably good to make it clear it's an example... maybe `example: /data` instead of e.g.?

spadgett commented:
> now that I see it, it's probably good to make it clear it's an example... maybe example: /data instead of e.g.?

Outside the scope of this PR, but...

I wonder if we should use placeholders in general because they make fields look like they're filled in. We're also inconsistent in how we use them: sometimes as a hint, example, or default value. Not sure how they work with screen readers as well.

Does Patternfly have any best practices for placeholders?

ajacobs21e commented:
@spadgett I'm checking to see if patternfly has guidance for placeholders. I don't see any instances where PF uses an example for a placeholder but I don't know if anything has been tested. 

OpenShift doesn't do a great job of distinguishing between an example for format/structure, a suggested example, or a default value. @fabianofranz I'll mock up some suggestions for this. @jwforres I'm thinking though that this might lead to a larger issue where we'll have to review all of our forms for consistency.

fabianofranz commented:
Removing WIP here, please review. 

Below are the latest screenshots. I addressed most of the comments, thanks everyone. A few things, to summarize:

- pvc's can be attached to both *deployment configs* or single *deployments* as we discussed
- in the attach page, note the message explaining what will happen when attaching to a deployment
- we didn't seem to have a conclusion about the attach URL, so I'm staying with the original proposal (`/project/:project/attachPersistentVolumeClaim`) which is similar to the one for create route. this is easy to change if we decide to
- like mentioned by @spadgett and @ajacobs21e we need to do better about placeholders and default values, but this is out of the scope of this PR
- I mentioned we need to improve how the volume mounts are displayed in the "Pod template" block versus the "Volumes" section, but this is out of the scope of this PR

**Browse/Storage (currently only lists pvc's)**

![storage-list](https://cloud.githubusercontent.com/assets/158611/11942550/8af3d0da-a81e-11e5-9c11-b98b60c4a13f.png)

**Browse/Storage/Detail**

![storage-item](https://cloud.githubusercontent.com/assets/158611/11942572/aa726336-a81e-11e5-95f5-0ac9804babba.png)

**"Attach storage" in the deployment config page**

![attach-dc](https://cloud.githubusercontent.com/assets/158611/11942620/07a8e886-a81f-11e5-9ec2-22822f116e11.png)

**Attaching to a deployment config**

![attach-storage-dc](https://cloud.githubusercontent.com/assets/158611/11942675/4e614f3e-a81f-11e5-9f79-ef6dc5e2688d.png)

**"Attach storage" in the single deployment page**

![attach-deployment](https://cloud.githubusercontent.com/assets/158611/11942630/11a64d74-a81f-11e5-98a0-a07f1b09de18.png)

**Attaching to a deployment**

![attach-storage-deployment](https://cloud.githubusercontent.com/assets/158611/11942655/3705340e-a81f-11e5-8733-591de6813bb0.png)

fabianofranz commented:
[test]

smarterclayton commented:
Visually I'm not sure I like volumes being presented as separate from the pod template. We should always make it clear they are part of the pod template.  Part of it is that the summary presentation here is insufficient to represent the pod completely - it's likely that that presentation is too abridged, and needs to be expanded visually so it's clear that pod -> {containers -> volume mounts, policies, volumes}, and I don't think we've achieved that yet.

smarterclayton commented:
The attach form looks good.  We can fiddle with volume name later - id say it's not important enough to take up this much space (or I would invert the ordering).

fabianofranz commented:
@smarterclayton I totally agree but I proposed to address the volumes/pod template visuals outside of the scope of this PR. It's like that today already, this PR only adds a link to the pvc where/the way it was already displayed. Should we try to address that here too?

smarterclayton commented:
As long as we have a follow up to address it in the future.

fabianofranz commented:
@smarterclayton ok, I Inverted the ordering and now mount path comes first, screens updated. Follow-up for the volumes visuals is https://github.com/openshift/origin/issues/6454.

jwforres commented:
@smarterclayton @fabianofranz Yes, agreed.

@spadgett and I were discussing the general issue of these pages yesterday,
the more we need to add to them the more cluttered and unbalanced they
get.  Pods, build configs, and deployment configs are the worst offenders,
most of the other pages dont have enough detail to be a problem yet.  We
can continue the discussion on #6454

On Mon, Dec 21, 2015 at 10:29 PM, Fabiano Franz <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> ok, I Inverted the
> ordering and now mount path comes first, screens updated. Follow-up for the
> volumes visuals is #6454 <https://github.com/openshift/origin/issues/6454>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6358#issuecomment-166494586>.
>

jwforres commented:
How about:
"The volume will be mounted into all containers. You can \<a>select specific containers\<\a\> instead."

jwforres commented:
Shouldn't you be able to Delete or Edit (raw) your PVC from the PVC's describe page?  It should be easy to add those assuming its possible, the actions are generic directives.

jwforres commented:
What does it look like if you attach to an RC that is not a deployment?

jwforres commented:
Do we need to help users as far as understanding that they need to redeploy (dep config case) or scale down / up or delete pods (deployment / RC cases) in order for the pods to have the mounts added.

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+ * Controller of the openshiftConsole
+ */
+angular.module('openshiftConsole')
+  .controller('AttachPersistentVolumeClaimController', function ($filter, $routeParams, $scope, $window, StorageService, DataService, Navigate, ProjectsService) {
```
reeeeally long controller name, I would be ok with using PVC here instead, honestly i'd be ok with PVC in the routes as well like @spadgett suggested

fabianofranz replied:
Ok, so `AttachPVCController` and `/attachPVC`, correct?

jwforres replied:
yep

On Tue, Dec 22, 2015 at 1:40 PM, Fabiano Franz <notifications@github.com>
wrote:

> In assets/app/scripts/controllers/attachPersistentVolumeClaim.js
> <https://github.com/openshift/origin/pull/6358#discussion_r48283504>:
>
> > @@ -0,0 +1,280 @@
> > +'use strict';
> > +
> > +/**
> > + * @ngdoc function
> > + * @name openshiftConsole.controller:CreateRouteController
> > + * @description
> > + * # CreateRouteController
> > + * Controller of the openshiftConsole
> > + */
> > +angular.module('openshiftConsole')
> > +  .controller('AttachPersistentVolumeClaimController', function ($filter, $routeParams, $scope, $window, StorageService, DataService, Navigate, ProjectsService) {
>
> Ok, so AttachPVCController and /attachPVC, correct?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6358/files#r48283504>.
>

spadgett replied:
Rename the file to `attachPVC.js` as well if you haven't.

fabianofranz replied:
fixed

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+        // check parameters validity
+        var validate = function() {
+          if ($scope.deploymentConfigName && $scope.deploymentName) {
+            sendError("Deployment and deployment config can't be both provided provided.");
```
this is OK for now, if we let you do the attach action from the PVC itself (which we should at some point) then we would provide a way to select a deployment config / deployment / RC when one wasnt passed as a param, see what we do for create route in certain cases

fabianofranz replied:
:+1:

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+          if (deploymentConfig) {
+            $scope.breadcrumbs.splice(2, 0, {
+              title: deploymentConfig.metadata.name,
+              link: "project/" + $routeParams.project + "/browse/deployments/" + deploymentConfig.metadata.name
```
don't build this link yourself, use the navigateResourceURL filter

fabianofranz replied:
fixed

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+
+          var deployment = $scope.attach.deployment;
+          if (deployment) {
+            var deploymentVersion = $filter("annotation")(deployment, "deploymentVersion");
```
can't assume an RC is a deployment, so deploymentVersion may not be there, even for an actual deployment (has dep config annotation) we can't assume the deploymentVersion annotation will always be there since anyone can remove annotations

fabianofranz replied:
I test `deploymentVersion` a couple line below (still addressing pure rc's though).

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+            });
+            var deploymentDeploymentConfigName = $filter("annotation")(deployment, "deploymentConfig");
+            if (deploymentDeploymentConfigName) {
+              $scope.breadcrumbs[2].link = "project/" + $routeParams.project + "/browse/deployments/" + deploymentDeploymentConfigName + "/" + deployment.metadata.name;
```
use navigateResourceURL

fabianofranz replied:
fixed

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+        var orderByDisplayName = $filter('orderByDisplayName');
+        var getErrorDetails = $filter('getErrorDetails');
+
+        if (!validate() || !load()) {
```
when would load ever return anything other than true? the error callbacks are all asynchronous

fabianofranz replied:
fixed

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+          if ($scope.attachPersistentVolumeClaimForm.$valid) {
+            // generate a volume name if not provided
+            if (!$scope.attach.volumeName) {
+              $scope.attach.volumeName = StorageService.generateSimpleName("volume-");
```
is there anything storage specific about generateSimpleName?  should it be a general filter that you pass a prefix and a length to?

fabianofranz replied:
moved to filter

spadgett replied:
Ideally this and ApplicationGenerator._generateSecret would go in the existing NameGenerator service with unit tests that verify their output matches the pattern we expect. Maybe a follow on

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+
+            // we must attach to at least one container (han-picked by user, or just "all")
+            if (!containerToAttachProvided()) {
+              sendError("You must specify at least one container.");
```
instead of sending this error to the top of the form, you should expand the containers list and have the error appear inline with that list

spadgett replied:
Agree with @jwforres. Can any of these errors be done with form validation before the submit button is clicked?

fabianofranz replied:
fixed

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+
+            // check the volume name wanted was not yet used in this pod template
+            if (!volumeNameIsUnique(name, podTemplate)) {
+              sendError("Volume named '" + name + "' already exists. Please choose another name.");
```
this should be a field error not a top level error

fabianofranz replied:
fixed

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+            if (mountPath) {
+              // if we want to mount, check if the mount path is unique in this pod template 
+              if (!volumeMountPathIsUnique(name, mountPath, podTemplate)) {
+                sendError("Volume mount in path '" + mountPath + "' already exists. Please choose another mount path.");
```
field level error

fabianofranz replied:
fixed

jwforres commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+              return;
+            }
+
+            if (mountPath) {
```
we are requiring you to have a container selected to mount to... but not requiring you to have a mountPath set (and not defaulting otherwise)? that's a bit confusing.  Are we making mounting it in optional, or not.

fabianofranz replied:
Pretty much the same is done in the cli, but I agree it's a bit confusing. Does it even make sense to attach a volume without forcing a mount path to be present? Could the user mount manually later if desired (and only in the containers you selected the volume would be available)?

jwforres commented on assets/app/scripts/controllers/storage.js:
```diff
@@ -0,0 +1,73 @@
+      .get($routeParams.project)
+      .then(_.spread(function(project, context) {
+        $scope.project = project;
+         watches.push(DataService.watch("persistentvolumeclaims", context, function(storage, action) {
```
replace storage with pvcs here

fabianofranz replied:
fixed

jwforres commented on assets/app/scripts/controllers/storage.js:
```diff
@@ -0,0 +1,73 @@
+  .controller('StorageController', function ($routeParams, $scope, AlertMessageService, DataService, ProjectsService, $filter, LabelFilter, Logger, $location, $anchorScroll) {
+    $scope.projectName = $routeParams.project;
+    $scope.pvcs = {};
+    $scope.unfilteredPvcs = {};
```
unfilteredPVCs  use upper case on abbreviations

fabianofranz replied:
fixed

jwforres commented on assets/app/scripts/filters/util.js:
```diff
@@ -117,6 +117,9 @@ angular.module('openshiftConsole')
           unit += (amount === "1" ? "core" : "cores");
           break;
+        case "storage":
+          unit = unit.substring(0, 1);
```
i'd expect this to behave like memory and add the "B" to it so you see GB instead of G, what units are actually being returned, since I see you are substring-ing it

fabianofranz replied:
Sometimes it's just `G` and sometimes it's `Gi` as referred by [some docs](https://docs.openshift.org/latest/dev_guide/persistent_volumes.html#requesting-storage). I'll keep substring-ing it but I'll add `B` like memory does, so that we have `GB`.

jwforres replied:
dont drop the i in Gi, it actually does mean something different :)

On Tue, Dec 22, 2015 at 3:11 PM, Fabiano Franz <notifications@github.com>
wrote:

> In assets/app/scripts/filters/util.js
> <https://github.com/openshift/origin/pull/6358#discussion_r48292989>:
>
> > @@ -117,6 +117,9 @@ angular.module('openshiftConsole')
> >            }
> >            unit += (amount === "1" ? "core" : "cores");
> >            break;
> > +        case "storage":
> > +          unit = unit.substring(0, 1);
>
> Sometimes it's just G and sometimes it's Gi as referred by some docs
> <https://docs.openshift.org/latest/dev_guide/persistent_volumes.html#requesting-storage>.
> I'll keep substring-ing it but I'll add B like memory does, so that we
> have GB.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6358/files#r48292989>.
>

fabianofranz replied:
fixed

spadgett replied:
Do we need `substring` at all? Any reason we can't reuse the memory case already defined above?

fabianofranz replied:
Ha, I wasn't aware we were using the gibitytes standards G (1000) and Gi (1024). Then I think we should reuse memory, we will end up having for example `GB` and `GiB`.

fabianofranz replied:
fixed to use the same rules as memory

jwforres commented on assets/app/scripts/filters/util.js:
```diff
@@ -341,4 +344,39 @@ angular.module('openshiftConsole')
+            accessModeString = longForm ? "RWX (Read-Write-Many)" : "Read-Write-Many";
+            break;
+        }
+        if (accessModeString && accessModes.indexOf(accessModeString) === -1) {
```
how but just always pushing accessModeString if its defined, and then doing one call to _.uniq at the end https://lodash.com/docs#uniq

fabianofranz replied:
fixed (I wasn't aware we had lodash) :muscle:

jwforres commented on assets/app/views/_volumes.html:
```diff
@@ -0,0 +1,50 @@
+  podTemplate
+  namespace
+ -->
+<div ng-repeat="volume in podTemplate.spec.volumes">
```
is there any reason to pass podTemplate, instead of just passing volumes?

fabianofranz replied:
fixed to pass the volumes array

jwforres commented on assets/app/views/_volumes.html:
```diff
@@ -0,0 +1,50 @@
+      <dt>Type:</dt>
+      <dd>persistent volume claim</dd>
+      <dt>Claim name:</dt>
+      <dd><a href="project/{{namespace}}/browse/persistentvolumeclaims/{{volume.persistentVolumeClaim.claimName}}">{{volume.persistentVolumeClaim.claimName}}</a></dd>
```
use navigateResourceURL filter

fabianofranz replied:
fixed

jwforres commented on assets/app/views/attachPersistentVolumeClaim.html:
```diff
@@ -0,0 +1,183 @@
+                <!-- attaching to a deployment config -->
+                <div class="form-group" ng-if="attach.deploymentConfig">
+                  <label for="deployment-config-name" class="required">Deployment config</label>
+                  <input
```
is this what we did on create route? @spadgett  was the service name always an input and disabled if it was pre-selected from URL params?

spadgett replied:
We hid the service input and updated the description at the top of the create route page to say, "Create a route to expose service {{serviceName}}."

jwforres commented on assets/app/views/attachPersistentVolumeClaim.html:
```diff
@@ -0,0 +1,183 @@
+                  <div>
+                    <span id="deployment-name-help" class="help-block">The pod templates of this deployment will specify volumes that attach to the persistent volume claim below. 
+                    <span ng-if="attach.deployment | annotation:'deploymentConfig'">
+                      Note that this will only apply to the pods of deployment <b>{{attach.deployment.metadata.name}}</b>, if you want volumes that apply for subsequent deployments then attach to the deployment config <b>{{attach.deployment | annotation:'deploymentConfig'}}</b> instead.</span>
```
this text is wrong if its an RC thats not a deployment

fabianofranz replied:
not shown if it's not a deployment

jwforres commented on assets/app/views/attachPersistentVolumeClaim.html:
```diff
@@ -0,0 +1,183 @@
+                  </div>
+                  <div class="has-error" ng-show="attachPersistentVolumeClaimForm.mountPath.$error.pattern && attachPersistentVolumeClaimForm.mountPath.$touched">
+                    <span class="help-block">
+                      Mount path must be a valid path to directory starting with <code>/</code>
```
"path to a directory"

fabianofranz replied:
fixed

jwforres commented on assets/app/views/browse/_pod-details.html:
```diff
@@ -48,52 +48,9 @@
-          </dl>
-        </div>    
+        <h4 style="margin-top: 20px;">Volumes</h4>
+        <a ng-if="pod | annotation:'deploymentConfig'" ng-href="project/{{project.metadata.name}}/attachPersistentVolumeClaim?deploymentconfig={{pod | annotation:'deploymentConfig'}}">Attach storage and redeploy</a>
```
is this what oc attach does for a pod? does it attach to the deploymentconfig that spawned the deployment that spawned the pod?

fabianofranz replied:
`oc` does not allow attaching to a pod at all, it fails with the message "error: may not update volume in pod <name> directly". Should we maybe do the same and just remove the link from the pod details page, to avoid confusion?

jwforres commented on assets/app/views/browse/persistent-volume-claim.html:
```diff
@@ -0,0 +1,51 @@
+                  <span ng-if="pvc.spec.resources.requests['storage']">
+                    {{pvc.spec.resources.requests['storage'] | usageWithUnits: 'storage'}}
+                  </span>
+                  <span ng-if="!pvc.spec.resources.requests['storage']">unknown</span>
```
I think this should be `<em>none</em>` instead of unknown, unknown to me implies that we failed to determine what the request was, whereas none implies that one was never set

fabianofranz replied:
fixed

spadgett commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+        $scope.project = project;
+
+        var containerToAttachProvided = function() {
+          for (var key in $scope.attach.containers.individual) {
```
Better to use `Object.keys()`.

spadgett commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+          return false;
+        };
+
+        var sendError = function(errorMessage, errorDetails) {
```
`displayError`?

fabianofranz replied:
fixed

spadgett commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+        };
+
+        var orderByDisplayName = $filter('orderByDisplayName');
+        var getErrorDetails = $filter('getErrorDetails');
```
Move these above where they're first called in case the DataService callback ever happens immediately.

fabianofranz replied:
fixed

spadgett commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+              }
+            }
+          }
+          return true;
```
Maybe just (untested)

```
return _.some(podTemplate.spec.volumes, 'name', name);
```

https://lodash.com/docs#some

spadgett commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+          return true;
+        };
+
+        var volumeNameIsUnique = function(name, podTemplate) {
```
`isVolumeNameUnique`?

fabianofranz replied:
actually I changed these to return `true` when exception: `volumeNameAlreadyUsed` and `volumeMountPathAlreadyUsed`, how does it sound?

spadgett replied:
Maybe it's my java background, but I like function names that start with `is`. Then it's obvious they're checking something and returning a boolean with no side effects.

jwforres replied:
agree, isVolumeNameUnique

On Tue, Dec 22, 2015 at 4:20 PM, Sam Padgett <notifications@github.com>
wrote:

> In assets/app/scripts/controllers/attachPersistentVolumeClaim.js
> <https://github.com/openshift/origin/pull/6358#discussion_r48299370>:
>
> > +                sendError("There are no persistent volumes in your project to attach.");
> > +              } else {
> > +                if (!$scope.attach.persistentVolumeClaim) {
> > +                  $scope.attach.persistentVolumeClaim = $scope.pvcs[0];
> > +                }
> > +              }
> > +            },
> > +            function(e) {
> > +              sendError("The persistent volume claims could not be loaded.", getErrorDetails(e));
> > +            }
> > +          );
> > +
> > +          return true;
> > +        };
> > +
> > +        var volumeNameIsUnique = function(name, podTemplate) {
>
> Maybe it's my java background, but I like function names that start with
> is. Then it's obvious they're checking something and returning a boolean
> with no side effects.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6358/files#r48299370>.
>

fabianofranz replied:
+1 java background, fixed :)

spadgett commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+          return true;
+        };
+
+        var volumeMountPathIsUnique = function(name, mountPath, podTemplate) {
```
`isVolumeMountPathUnique`?

spadgett commented on assets/app/scripts/controllers/attachPersistentVolumeClaim.js:
```diff
@@ -0,0 +1,280 @@
+          if (!$scope.deploymentConfigName && !$scope.deploymentName) {
+            sendError("A deployment or deployment config must be provided.");
+            return false;
+          }
```
These errors only occur when the query parameters are incorrect, right? So I manually changed a URL or we have a bug?

We might use `Navigate.toErrorPage()` instead like we do in CreateFromImageController.

spadgett replied:
> These errors only occur when the query parameters are incorrect, right?

If true, I'd change the name to something more specific like `validateRouteParams` to avoid confusion with form validation and perhaps check the route params directly at the top of the file before assigning scope values.

fabianofranz replied:
Good advice, I changed the func name and using navigate now, also removed the boolean return value from the function.

spadgett commented on assets/app/views/storage.html:
```diff
@@ -0,0 +1,46 @@
+        </tbody>
+        <tbody ng-repeat="pvc in pvcs | orderObjectsByDate : true">
+          <tr>
+            <td data-title="Name"><a href="{{pvc | navigateResourceURL}}">{{pvc.metadata.name}}</a></td>
```
`ng-href`

fabianofranz replied:
fixed

spadgett commented on assets/app/scripts/controllers/storage.js:
```diff
@@ -0,0 +1,73 @@
+          if (!action && $location.hash()) {
+            // Wait until the digest loop completes.
+            setTimeout($anchorScroll, 10);
+          }
```
Is this being used? I didn't see any anchors in storage.html.

fabianofranz replied:
good catch, removed

spadgett commented on assets/app/scripts/filters/util.js:
```diff
@@ -341,4 +344,39 @@ angular.module('openshiftConsole')
+          case "ReadWriteMany":
+            accessModeString = longForm ? "RWX (Read-Write-Many)" : "Read-Write-Many";
+            break;
+        }
```
Should we at least log a warning if we see a value we don't recognize?

fabianofranz replied:
I'll actually change to use the value as-is if it's not one of the ones we recognize.

spadgett replied:
Yeah, sounds better.

spadgett commented on assets/app/scripts/services/storage.js:
```diff
@@ -0,0 +1,35 @@
+        length = 5;
+      }
+      var randomString = Math.round((Math.pow(36, length + 1) - Math.random() * Math.pow(36, length))).toString(36).slice(1);
+      return prefix + randomString;
```
We have similar code in `ApplicationGenerator._generateSecret`. Is it possible to reuse that?

fabianofranz replied:
This is actually different, it's not a GUID but a 5-letters random that will give you for example `volume-7a91y`. This is how the CLI does it so I'd like to keep the same approach so we have similar id's not matter where the volume were attached. Notice as per @jwforres suggestion we moved this to a filter.

spadgett commented on assets/app/scripts/services/storage.js:
```diff
@@ -0,0 +1,35 @@
+  .factory("StorageService", function() {
+    function StorageService() {}
+
+    StorageService.prototype.generateSimpleName = function(prefix, length) {
```
I prefer something like this to new/prototype.

```
angular.module("openshiftConsole")
 .factory("StorageService", function() {
    return {
      generateSimpleName: function(prefix, length) { /* ... */ },
      createVolume: function(name, persistentVolumeClaim) { /* ... */ }  
      // ...
    };
  });
```

fabianofranz replied:
+1, fixed

benjaminapetersen replied:
agree, nice!

spadgett commented on assets/app/views/_volumes.html:
```diff
@@ -0,0 +1,50 @@
+      <dt ng-if-start="volume.gitRepo.revision">Revision:</dt>
+      <dd ng-if-end>{{volume.gitRepo.revision}}</dd>
+    </div>
+  </dl>
```
Is it ever possible all these ng-if's are false and we have a blank page? :)

spadgett commented on assets/app/views/attachPersistentVolumeClaim.html:
```diff
@@ -0,0 +1,198 @@
+                            ng-value="pvc"
+                            aria-describedby="pvc-help">
+                        </td>
+                        <td><a href="{{pvc | navigateResourceURL}}">{{pvc.metadata.name}}</a></td>
```
`ng-href`

fabianofranz replied:
fixed

spadgett commented on assets/app/views/attachPersistentVolumeClaim.html:
```diff
@@ -0,0 +1,198 @@
+                      ng-click="attachPersistentVolumeClaim()"
+                      ng-disabled="attachPersistentVolumeClaimForm.$invalid || disableInputs || !attachPersistentVolumeClaim || !containerToAttachProvided()"
+                      value="">Attach</button>
+                  <a class="btn btn-default btn-lg" href="#" back>Cancel</a>
```
`role=button`

spadgett replied:
I don't think this will be disabled with the fieldset since it's a link.

fabianofranz replied:
I doesn't, but I think it's desired since so that you can go back, isn't it?

spadgett replied:
I was thinking about someone clicking Submit then Cancel while we're still processing the form, but maybe not a problem in practice.

spadgett commented on assets/app/views/attachPersistentVolumeClaim.html:
```diff
@@ -0,0 +1,198 @@
+      <breadcrumbs breadcrumbs="breadcrumbs"></breadcrumbs>
+      <alerts alerts="alerts"></alerts>
+      <div class="row">
+        <div class="create-route-icon col-md-2 gutter-top hidden-sm hidden-xs">
```
Rename the `create-route-icon` class if we're going to use it other places.

fabianofranz replied:
fixed

spadgett commented on assets/app/views/browse/_pod-details.html:
```diff
@@ -48,52 +48,9 @@
-          </dl>
-        </div>    
+        <h4 style="margin-top: 20px;">Volumes</h4>
+        <a ng-if="pod | annotation:'deploymentConfig'" ng-href="project/{{project.metadata.name}}/attachPVC?deploymentconfig={{pod | annotation:'deploymentConfig'}}">Attach storage and redeploy</a>
```
Make sure the `deploymentconfig` query parameter is properly encoded. The annotation could in theory contain any value since it's editable.

spadgett commented on assets/app/views/storage.html:
```diff
@@ -0,0 +1,46 @@
+          </tr>
+        </thead>
+        <tbody ng-if="(pvcs | hashSize) == 0">
+          <tr><td colspan="7"><em>{{emptyMessage}}</em></td></tr>
```
`colspan="5"`?

fabianofranz replied:
correct, missed that. fixed

spadgett commented on assets/app/views/attachPVC.html:
```diff
@@ -0,0 +1,198 @@
+                      class="btn btn-primary btn-lg"
+                      ng-click="attachPVC()"
+                      ng-disabled="attachPVC.$invalid || disableInputs || !attachPVC || !containerToAttachProvided()"
+                      value="">Attach</button>
```
Is `value=""` needed?

fabianofranz replied:
nope, fixed

spadgett commented on assets/app/views/attachPVC.html:
```diff
@@ -0,0 +1,198 @@
+                      autocorrect="off"
+                      autocapitalize="off"
+                      spellcheck="false"
+                      disabled
```
Is this input and the one above always disabled? Any reason to not just use a static text?

http://getbootstrap.com/css/#forms-controls-static

fabianofranz replied:
Always disabled until we have "attach to..." in future. Doesn't look very good as static text, what do you think?

![screenshot from 2015-12-22 20-12-31](https://cloud.githubusercontent.com/assets/158611/11967241/61e84ef8-a8e8-11e5-85a3-d2c819035a57.png)

spadgett replied:
I think I prefer static text to a disabled input, but hard to say without seeing the whole form.

We could also do what we did for create route: https://github.com/openshift/origin/pull/6358#discussion_r48297410

fabianofranz replied:
Sounds good, I fixed to how it's done in create route. It says "Attach an existing persistent volume claim to the template of deployment config {{name}}", same for when we only have a deployment or rc.

fabianofranz commented:
Ok, thanks folks for the comments, I addressed all/most of them. Thinks I didn't answer inline:

- @jwforres works fine for RC's that are not deployments too, I'll also make sure QE add a test for that since we don't have it yet. 
- @jwforres when attaching to a deployment config you don't need to redeploy manually, attaching will actually trigger it. we could add some info for when it's deployment/rc, maybe add [here](https://github.com/openshift/origin/pull/6358/files#diff-d5dee70e05017b2b8915d930d5521255R51)?
- @jwforres added raw edit and delete.

What else?

spadgett commented:
This error seems a little odd to me. There's no guidance on how to fix. I think we need to at least point to the documentation on how to create a persistent volume.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12123695/6f172f86-b3b0-11e5-8ed6-384454e9e9e1.png)

For comparison, here is what we show when you click Add to Project and there are no templates or image streams. We could reuse this style.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12123747/af5e8364-b3b0-11e5-9621-715c30abfcfc.png)

https://github.com/openshift/origin/blob/master/assets/app/views/create.html#L12-L27

spadgett commented:
For succeeded pods, we show a Volumes section with no content.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12123881/5e7c5ea2-b3b1-11e5-8e2f-23ef5af854e7.png)

spadgett commented on assets/app/scripts/controllers/attachPVC.js:
```diff
@@ -0,0 +1,277 @@
+            function(pvcs) {
+              $scope.pvcs = orderByDisplayName(pvcs.by("metadata.name"));
+              if (!$scope.pvcs.length) {
+                displayError("There are no persistent volumes in your project to attach.");
```
"There are no persistent volume claims in your project to attach." ?

fabianofranz replied:
fixed

fabianofranz commented:
@spadgett good advice about the "no pvcs" scenario, I replaced that alert with a page similar to Add to Project. Just note that I'm pointing to the docs instead of displaying an `oc` command, since creating a pvc not attached to anything is a little more complex (must be done from a template).

![openshift web console](https://cloud.githubusercontent.com/assets/158611/12153485/3082a02e-b4a1-11e5-9fc5-49ebb004c173.png)

Now about the empty Volumes section I couldn't reproduce, my succeeded "ruby-sample-build-xx-build" pods are displaying their volumes correctly. Could you export the template of yours, or any other info I could use to reproduce?

spadgett commented:
> Now about the empty Volumes section I couldn't reproduce, my succeeded "ruby-sample-build-xx-build" pods are displaying their volumes correctly. Could you export the template of yours, or any other info I could use to reproduce?

I just pulled down the latest, and I'm no longer seeing it.

spadgett commented on assets/app/views/attachPVC.html:
```diff
@@ -0,0 +1,172 @@
+          </p>
+
+          <p ng-if="attach.deploymentConfig"><a href="{{attach.deploymentConfig | navigateResourceURL}}">Back to deployment config</a></p>
+          <p ng-if="attach.deployment"><a href="{{attach.deployment | navigateResourceURL}}">Back to deployment</a></p>
```
This is a nit, but if you got here from the browse page page, the link will take you back to a deployment config, not the pod.

spadgett commented:
LGTM

[merge]
